stdlib/either.mli

```diff
@@ -0,0 +1,88 @@
+(**************************************************************************)
+(*                                                                        *)
+(*                                 OCaml                                  *)
+(*                                                                        *)
+(*         Gabriel Scherer, projet Parsial, INRIA Saclay                  *)
+(*                                                                        *)
+(*   Copyright 2019 Institut National de Recherche en Informatique et     *)
+(*     en Automatique.                                                    *)
+(*                                                                        *)
+(*   All rights reserved.  This file is distributed under the terms of    *)
+(*   the GNU Lesser General Public License version 2.1, with the          *)
+(*   special exception on linking described in the file LICENSE.          *)
+(*                                                                        *)
+(**************************************************************************)
+
+(** Either type.
+
+    Either is the simplest and most generic sum type. It is isomorphic
+    to the [result] type; it should be used when there is no implied
+    success/failure semantics in the choice of one of the constructors.
+
+    @since 4.11 *)
+
+(* Unlike [result], no [either] type is made available in Stdlib,
+   one needs to access [Either.t] explicitly:
+
+   - This type is less common in typical OCaml codebases,
+     which prefer domain-specific variant types whose constructors
+     carry more meaning.
+   - Adding this to Stdlib would raise warnings in existing codebases
+     that already use a constructor named Left or Right:
+     + when opening a module that exports such a name,
+       warning 45 is raised
+     + adding a second constructor of the same name in scope kicks
+       in the disambiguation mechanisms, and warning 41 may now
+       be raised by existing code.
+
+   If the use becomes more common in the future we can always
+   revisit this choice.
+*)
+
+type ('a, 'b) t = Left of 'a | Right of 'b (**)
+(** A value of [('a, 'b) Either.t] contains either a value of ['a]  or a value
+    of ['b] *)
+
+val left : 'a -> ('a, 'b) t
+(** [left v] is [Left v]. *)
+
+val right : 'b -> ('a, 'b) t
+(** [right v] is [Right v]. *)
+
+val get_left : ('a, 'b) t -> 'a option
+(** [get_left (Left v)] is [Some v], [get_left (Right _)] is [None] *)
```

Comment:
The behaviours for these names are inconsistent with `Result.get_{ok,error}` and `Option.get`. `get` is for when you know the thing is here (and raises `Invalid_argument` if it's not). I would like these names with that semantics. Names and signature for what you suggest consistent with `Result` and `Option` could be:
```ocaml
val left_value : ('a, 'b) t -> default:'a -> 'a 
val right_value : ('a, 'b) t -> default:'b -> 'b
```
or if you want to keep the current signatures:
```ocaml
val find_left : ('a, 'b) t -> 'a option 
val find_right : ('a, 'b) t -> 'b option
```

Reply:
Sure, I'll use `find_`. I'm not so interested in the more complex accessors, I think the `option` result is useful to be combined with option-to-{list,seq} transformations.

Reply:
I vote for the `find_left` and `find_right` version.

Reply:
I would still like to have the `get_{left,right}` accessors for 1) consistency with `Result` and `Option` 2) the error is slightly more informative than `function Left v -> v | _ -> assert false` when it fails.